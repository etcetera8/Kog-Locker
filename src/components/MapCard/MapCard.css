.map-card {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center; }

#map {
  margin-top: 5px;
  margin: 0 auto; }

.target-name {
  color: #fff;
  margin: 8px;
  font-size: 1em; }

.stats {
  margin: 0; }
