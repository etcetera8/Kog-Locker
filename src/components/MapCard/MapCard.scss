$orange: rgba(255,100,0,1);

.map-card {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

#map {
  margin-top: 5px;
  margin: 0 auto;
}

.target-name {
  color: #fff;
  margin: 8px;
  font-size: .9em;
}

.stats {
  margin: 0;
}

.goal {
  margin: 0px;
  display: inline;
}

.form {
  margin: 0;
  display: block;
  text-align: center;
}

.new-goal {
  background: none;
  border: 1px solid $orange;
  color: #fff;
  margin: 0 0 0 12px;
  width: 35px;
}