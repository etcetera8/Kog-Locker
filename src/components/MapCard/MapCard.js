import React, { Component } from 'react';
import { connect } from 'react-redux';
import ReactDOM from 'react-dom';
import { addMap, addNewGoal } from '../../actions/actionIndex';
import './MapCard.css'
import loadingGif from '../../assets/Spinner-1s-144px.gif'
import PropTypes from 'prop-types';

export class MapCard extends Component {
  constructor() {
    super()
    this.state = {
      goalTime: 0,
      newTime:0,
      loading: true
    }
  }

  async componentDidUpdate() {
    if (!this.props.targetMap) {
      await this.loadMap(this.props.userTarget.polyline, this.props.userTarget.end_latlng);
    }
  }

  componentDidMount() {
    if (this.props.targetMap && this.props.userTarget) {
      const {polyline, athlete_segment_stats} = this.props.userTarget;
      const node = ReactDOM.findDOMNode(this.refs.map);
      const mapConfig = this.initialMap()
      const decodedPolyline = this.decodePolyline(polyline);
      decodedPolyline.setMap(mapConfig)
      const goalTime = athlete_segment_stats ? ((athlete_segment_stats.pr_elapsed_time / 60)-((athlete_segment_stats.pr_elapsed_time / 60) * 0.03)).toFixed(2) : 0;
      this.setState({goalTime, loading: false})
    }
  }

  initialMap() {
    const lat = this.props.userTarget.end_latlng[0];
    const lng = this.props.userTarget.end_latlng[1];
    const node = ReactDOM.findDOMNode(this.refs.map);
    const mapConfig = new this.props.google.maps.Map(node, {
      zoom: 12,
      center: {lat, lng},
      mapTypeId: 'terrain'
    })
    return mapConfig
  }

  decodePolyline(targetPolyline) {
    const decodedSets = this.props.google.maps.geometry.encoding.decodePath(targetPolyline);
    const array = decodedSets.map(coord => JSON.stringify(coord));
    const parsedArray = array.map(coord => JSON.parse(coord));
    this.props.addMap(parsedArray)

    const polyline = new this.props.google.maps.Polyline({
      path: parsedArray,
      strokeColor: '#FF0000',
      strokeOpacity: 0.8,
      strokeWeight: 2,
      fillColor: '#FF0000',
      fillOpacity: 0
    })
    return polyline;
  }

  loadMap = (targetPolyline, latLong) => {
    if (this.props.userTarget && this.props.google.maps.geometry.encoding) {
      const { athlete_segment_stats } = this.props.userTarget;
      const { google } = this.props;
      const mapRef = this.refs.map;
      const node = ReactDOM.findDOMNode(mapRef);

      const mapConfig = this.initialMap();
      const polyline = this.decodePolyline(targetPolyline);

      polyline.setMap(mapConfig);
      this.map = new google.maps.Map(node, mapConfig);

      const goalTime = athlete_segment_stats ? ((athlete_segment_stats.pr_elapsed_time / 60)-((athlete_segment_stats.pr_elapsed_time / 60) * 0.03)).toFixed(2) : 0;
      this.setState({goalTime, loading: false});
    }

  }

  handleChange = (e) => {
    this.setState({newTime: e.target.value})
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const newGoal = this.state.newTime
    this.props.userTarget.athlete_segment_stats.goalTime = parseInt(this.state.newTime)
    this.props.addNewGoal(this.props.userTarget);
    this.setState({newTime: 0})
  }

  render() {
    const {name, athlete_segment_stats, goalTime} = this.props.userTarget
    const style = {
      width: '200px',
      height: '200px',
    }

    return (
      <section className='card map-card'>
        <h2 className='title'>Target</h2>
        {this.state.loading && 
          <div><img src={loadingGif}/></div>
        }
        { athlete_segment_stats &&
          <div>
            <h3 className="target-name">{name}</h3>
            <div id="map" className='map' ref='map' style={style}>
            </div>
            <span className='card-data goal'>Goal Time: <span className='nums'>{athlete_segment_stats.goalTime.toFixed(2)} mins</span></span>
            <form className="form card-data"> Update Goal:
              <input
                className="new-goal"
                type="number"
                value={this.state.newTime}
                onChange={this.handleChange}
              />
              <button 
                className="new-segment-button" 
                onClick={this.handleSubmit}
              >
                <i className="fas fa-plus"></i>
              </button>
            </form>
            <span className='card-data stats'>Best Time: <span className='nums'>{(athlete_segment_stats.pr_elapsed_time / 60).toFixed(2)} mins</span></span>
          </div>
        }
      </section>  
    )
  }
};

const mapStateToProps = (state) => ({
  targetMap: state.targetMap,
  userTarget: state.userTarget
})

const mapDispatchToProps = (dispatch) => ({
  addMap: (map) => dispatch(addMap(map)),
  addNewGoal: goal => dispatch(addNewGoal(goal))
})

MapCard.propTypes = {
  userTarget: PropTypes.object.isRequired,
  google: PropTypes.object.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(MapCard)
