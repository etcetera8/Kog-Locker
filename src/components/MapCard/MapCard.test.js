import React from 'react';
import { shallow, mount } from 'enzyme';
import {MapCard} from './MapCard';

describe('MapCard' , () => {
  let wrapper;
  
  beforeEach( () => {
    const mockProps = {name: 'parker', athlete_segment_stats: {stats: 'hi'}, end_latlng: [0, 1] }
    const mockTarget = {name: "road", polyline: '12345', end_latlng: [0, 1] }
    wrapper = shallow(<MapCard 
                        userTarget={mockProps}
                        google={{maps:{}}}
                      />);
  })

  it('should have a default state', () => {
    const defaultState = {"goalTime": 0, "loading": true, "newTime": 0}
    expect(wrapper.state()).toEqual(defaultState)
  })

  it('the handleChange function shoud update the state when fired' , () => {
    expect(wrapper.state().newTime).toEqual(0)
    wrapper.find('input').first().simulate('change', {target:{value: 7}})
    expect(wrapper.state().newTime).toEqual(7)
  })

  it('the handlesubmit function should update the state when fired', () => {
    const mockEvent = {preventDefault: jest.fn()}
    expect(wrapper.state().goalTime).toEqual(0)
    wrapper.state().newTime = 7;
    wrapper.find('button').first().simulate('click', mockEvent)
    expect(wrapper.state().goalTime).toEqual(7)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

})