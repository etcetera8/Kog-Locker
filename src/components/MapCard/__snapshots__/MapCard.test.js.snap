// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapCard should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MapCard
    google={
        Object {
            "maps": Object {},
          }
    }
    userTarget={
        Object {
            "athlete_segment_stats": Object {
              "stats": "hi",
            },
            "end_latlng": Array [
              0,
              1,
            ],
            "name": "parker",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="title"
>
          Target
</h2>,
        <div>
          <img
                    src="Spinner-1s-144px.gif"
          />
</div>,
        <div>
          <h3
                    className="target-name"
          >
                    parker
          </h3>
          <div
                    className="map"
                    id="map"
                    style={
                              Object {
                                        "height": "200px",
                                        "width": "200px",
                                      }
                    }
          />
          <span
                    className="card-data goal"
          >
                    Goal Time: 
                    <span
                              className="nums"
                    >
                              0
                               mins
                    </span>
          </span>
          <form
                    className="form"
          >
                    <input
                              className="new-goal"
                              onChange={[Function]}
                              placeholder="Update goal time"
                              type="number"
                              value={0}
                    />
                    <button
                              className="new-segment-button"
                              onClick={[Function]}
                    >
                              <i
                                        className="fas fa-plus"
                              />
                    </button>
          </form>
          <span
                    className="card-data stats"
          >
                    Best Time: 
                    <span
                              className="nums"
                    >
                              NaN
                               mins
                    </span>
          </span>
</div>,
      ],
      "className": "card map-card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Target",
          "className": "title",
        },
        "ref": null,
        "rendered": "Target",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            src="Spinner-1s-144px.gif"
/>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "src": "Spinner-1s-144px.gif",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3
              className="target-name"
>
              parker
</h3>,
            <div
              className="map"
              id="map"
              style={
                            Object {
                                          "height": "200px",
                                          "width": "200px",
                                        }
              }
/>,
            <span
              className="card-data goal"
>
              Goal Time: 
              <span
                            className="nums"
              >
                            0
                             mins
              </span>
</span>,
            <form
              className="form"
>
              <input
                            className="new-goal"
                            onChange={[Function]}
                            placeholder="Update goal time"
                            type="number"
                            value={0}
              />
              <button
                            className="new-segment-button"
                            onClick={[Function]}
              >
                            <i
                                          className="fas fa-plus"
                            />
              </button>
</form>,
            <span
              className="card-data stats"
>
              Best Time: 
              <span
                            className="nums"
              >
                            NaN
                             mins
              </span>
</span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "parker",
              "className": "target-name",
            },
            "ref": null,
            "rendered": "parker",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "map",
              "id": "map",
              "style": Object {
                "height": "200px",
                "width": "200px",
              },
            },
            "ref": "map",
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Goal Time: ",
                <span
                  className="nums"
>
                  0
                   mins
</span>,
              ],
              "className": "card-data goal",
            },
            "ref": null,
            "rendered": Array [
              "Goal Time: ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    0,
                    " mins",
                  ],
                  "className": "nums",
                },
                "ref": null,
                "rendered": Array [
                  0,
                  " mins",
                ],
                "type": "span",
              },
            ],
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  className="new-goal"
                  onChange={[Function]}
                  placeholder="Update goal time"
                  type="number"
                  value={0}
/>,
                <button
                  className="new-segment-button"
                  onClick={[Function]}
>
                  <i
                                    className="fas fa-plus"
                  />
</button>,
              ],
              "className": "form",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "new-goal",
                  "onChange": [Function],
                  "placeholder": "Update goal time",
                  "type": "number",
                  "value": 0,
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <i
                    className="fas fa-plus"
/>,
                  "className": "new-segment-button",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "fas fa-plus",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
                "type": "button",
              },
            ],
            "type": "form",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Best Time: ",
                <span
                  className="nums"
>
                  NaN
                   mins
</span>,
              ],
              "className": "card-data stats",
            },
            "ref": null,
            "rendered": Array [
              "Best Time: ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "NaN",
                    " mins",
                  ],
                  "className": "nums",
                },
                "ref": null,
                "rendered": Array [
                  "NaN",
                  " mins",
                ],
                "type": "span",
              },
            ],
            "type": "span",
          },
        ],
        "type": "div",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="title"
>
            Target
</h2>,
          <div>
            <img
                        src="Spinner-1s-144px.gif"
            />
</div>,
          <div>
            <h3
                        className="target-name"
            >
                        parker
            </h3>
            <div
                        className="map"
                        id="map"
                        style={
                                    Object {
                                                "height": "200px",
                                                "width": "200px",
                                              }
                        }
            />
            <span
                        className="card-data goal"
            >
                        Goal Time: 
                        <span
                                    className="nums"
                        >
                                    0
                                     mins
                        </span>
            </span>
            <form
                        className="form"
            >
                        <input
                                    className="new-goal"
                                    onChange={[Function]}
                                    placeholder="Update goal time"
                                    type="number"
                                    value={0}
                        />
                        <button
                                    className="new-segment-button"
                                    onClick={[Function]}
                        >
                                    <i
                                                className="fas fa-plus"
                                    />
                        </button>
            </form>
            <span
                        className="card-data stats"
            >
                        Best Time: 
                        <span
                                    className="nums"
                        >
                                    NaN
                                     mins
                        </span>
            </span>
</div>,
        ],
        "className": "card map-card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Target",
            "className": "title",
          },
          "ref": null,
          "rendered": "Target",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              src="Spinner-1s-144px.gif"
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "src": "Spinner-1s-144px.gif",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3
                className="target-name"
>
                parker
</h3>,
              <div
                className="map"
                id="map"
                style={
                                Object {
                                                "height": "200px",
                                                "width": "200px",
                                              }
                }
/>,
              <span
                className="card-data goal"
>
                Goal Time: 
                <span
                                className="nums"
                >
                                0
                                 mins
                </span>
</span>,
              <form
                className="form"
>
                <input
                                className="new-goal"
                                onChange={[Function]}
                                placeholder="Update goal time"
                                type="number"
                                value={0}
                />
                <button
                                className="new-segment-button"
                                onClick={[Function]}
                >
                                <i
                                                className="fas fa-plus"
                                />
                </button>
</form>,
              <span
                className="card-data stats"
>
                Best Time: 
                <span
                                className="nums"
                >
                                NaN
                                 mins
                </span>
</span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "parker",
                "className": "target-name",
              },
              "ref": null,
              "rendered": "parker",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "map",
                "id": "map",
                "style": Object {
                  "height": "200px",
                  "width": "200px",
                },
              },
              "ref": "map",
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Goal Time: ",
                  <span
                    className="nums"
>
                    0
                     mins
</span>,
                ],
                "className": "card-data goal",
              },
              "ref": null,
              "rendered": Array [
                "Goal Time: ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      0,
                      " mins",
                    ],
                    "className": "nums",
                  },
                  "ref": null,
                  "rendered": Array [
                    0,
                    " mins",
                  ],
                  "type": "span",
                },
              ],
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    className="new-goal"
                    onChange={[Function]}
                    placeholder="Update goal time"
                    type="number"
                    value={0}
/>,
                  <button
                    className="new-segment-button"
                    onClick={[Function]}
>
                    <i
                                        className="fas fa-plus"
                    />
</button>,
                ],
                "className": "form",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "new-goal",
                    "onChange": [Function],
                    "placeholder": "Update goal time",
                    "type": "number",
                    "value": 0,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <i
                      className="fas fa-plus"
/>,
                    "className": "new-segment-button",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "fas fa-plus",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "i",
                  },
                  "type": "button",
                },
              ],
              "type": "form",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Best Time: ",
                  <span
                    className="nums"
>
                    NaN
                     mins
</span>,
                ],
                "className": "card-data stats",
              },
              "ref": null,
              "rendered": Array [
                "Best Time: ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "NaN",
                      " mins",
                    ],
                    "className": "nums",
                  },
                  "ref": null,
                  "rendered": Array [
                    "NaN",
                    " mins",
                  ],
                  "type": "span",
                },
              ],
              "type": "span",
            },
          ],
          "type": "div",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
